src/arena.rs

```diff
@@ -720,6 +842,67 @@ where
         }
         list.iter(arena)
     }
+
+    /// Ensures that both deques are stored in the same direction.  It doesn't matter _which_
+    /// direction, as long as they're the same, so do the minimum amount of work to bring this
+    /// about.  (In particular, if we've already calculated the reversal of one of the deques,
+    /// reverse that one.)
+    fn ensure_same_direction(&mut self, arena: &mut DequeArena<T>, other: &mut Deque<T>) {
+        if self.direction == other.direction {
+            return;
+        }
+        if self.list.have_reversal(arena) {
+            self.list.reverse(arena);
+            self.direction = !self.direction;
+        } else {
+            other.list.reverse(arena);
+            other.direction = !other.direction;
+        }
+    }
+}
+
+impl<T> Deque<T>
+where
+    T: Clone,
+{
+    pub fn equals_with<F>(mut self, arena: &mut DequeArena<T>, mut other: Deque<T>, eq: F) -> bool
+    where
+        F: FnMut(&T, &T) -> bool,
+    {
+        self.ensure_same_direction(arena, &mut other);
+        self.list.equals_with(arena, other.list, eq)
+    }
+
+    pub fn cmp_with<F>(
+        mut self,
+        arena: &mut DequeArena<T>,
+        mut other: Deque<T>,
+        cmp: F,
+    ) -> std::cmp::Ordering
+    where
+        F: FnMut(&T, &T) -> std::cmp::Ordering,
+    {
+        self.ensure_same_direction(arena, &mut other);
+        self.list.cmp_with(arena, other.list, cmp)
```

Comment:
Hmm, now that I look at this again, I'm not sure if this is right for `cmp`, since lexicographic comparison needs to iterate from the front of each deque.

Reply:
Added an extra test case and a fix for this